physical/s3.go

```diff
@@ -0,0 +1,207 @@
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/aws/credentials"
+	"github.com/awslabs/aws-sdk-go/service/s3"
+)
+
+// S3Backend is a physical backend that stores data
+// within an S3 bucket.
+type S3Backend struct {
+	bucket string
+	client *s3.S3
+	l      sync.Mutex
+}
+
+// newS3Backend constructs a S3 backend using a pre-existing
+// bucket. Credentials can be provided to the backend, sourced
+// from the environment, AWS credential files or by IAM role.
+func newS3Backend(conf map[string]string) (Backend, error) {
+
+	bucket, ok := conf["bucket"]
+	if !ok {
+		return nil, fmt.Errorf("'bucket' must be set")
+	}
+
+	access_key, ok := conf["access_key"]
+	if !ok {
+		access_key = ""
+	}
+	secret_key, ok := conf["secret_key"]
+	if !ok {
+		secret_key = ""
+	}
+	region, ok := conf["region"]
+	if !ok {
+		region = os.Getenv("AWS_DEFAULT_REGION")
+		if region == "" {
+			region = "us-east-1"
+		}
+	}
+
+	creds := credentials.NewChainCredentials([]credentials.Provider{
+		&credentials.StaticProvider{Value: credentials.Value{
+			AccessKeyID:     access_key,
+			SecretAccessKey: secret_key,
+		}},
+		&credentials.EnvProvider{},
+		&credentials.SharedCredentialsProvider{Filename: "", Profile: ""},
+		&credentials.EC2RoleProvider{},
+	})
+
+	s3conn := s3.New(&aws.Config{
+		Credentials: creds,
+		Region:      region,
+	})
+
+	_, err := s3conn.HeadBucket(&s3.HeadBucketInput{Bucket: &bucket})
+	if err != nil {
+		return nil, fmt.Errorf("unable to access bucket '%s': %v", bucket, err)
+	}
+
+	s := &S3Backend{
+		client: s3conn,
+		bucket: bucket,
+	}
+	return s, nil
+}
+
+// Put is used to insert or update an entry
+func (s *S3Backend) Put(entry *Entry) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+	defer metrics.MeasureSince([]string{"s3", "put"}, time.Now())
+
+	_, err := s.client.PutObject(&s3.PutObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(entry.Key),
+		Body:   bytes.NewReader(entry.Value),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Get is used to fetch an entry
+func (s *S3Backend) Get(key string) (*Entry, error) {
+	s.l.Lock()
+	defer s.l.Unlock()
+	defer metrics.MeasureSince([]string{"s3", "get"}, time.Now())
+
+	resp, err := s.client.GetObject(&s3.GetObjectInput{
+		Bucket: aws.String(s.bucket),
+		Key:    aws.String(key),
+	})
+
+	if awserr := aws.Error(err); awserr != nil {
+		// Return nil on 404s, error on anything else
+		if awserr.StatusCode == 404 {
+			return nil, nil
+		} else {
+			return nil, err
+		}
+	}
+
+	data := make([]byte, *resp.ContentLength)
+	_, err = io.ReadFull(resp.Body, data)
+	if err != nil {
+		return nil, err
+	}
+
+	ent := &Entry{
+		Key:   key,
+		Value: data,
+	}
+
+	return ent, nil
+}
+
+// Delete is used to permanently delete an entry
+func (s *S3Backend) Delete(key string) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+	defer metrics.MeasureSince([]string{"s3", "delete"}, time.Now())
+
+	listResp, err := s.client.ListObjects(&s3.ListObjectsInput{
```

Comment:
Delete will name only a single object, so you should be able to do a `DeleteObject` against the exact named key.